source/features/useful-not-found-page.tsx

```diff
@@ -29,71 +27,247 @@ async function checkAnchor(anchor: HTMLAnchorElement): Promise<void> {
+	if (parts[2] === 'blob') {
 		parts[2] = 'tree';
 	}
 
 	return parts;
 }
 
-// If the resource was deleted, link to the commit history
-async function addCommitHistoryLink(bar: Element): Promise<void> {
-	const parts = parseCurrentURL();
-	parts[2] = 'commits';
-	const url = '/' + parts.join('/');
-	if (await is404(location.origin + url)) {
+async function getUrlToFileOnDefaultBranch(): Promise<string | undefined> {
+	// Get the current branch
+	const url = new GitHubURL(location.href);
+	const currentBranch = url.branch;
+
+	if (!currentBranch) {
 		return;
 	}
 
-	bar.after(
-		<p className="container mt-4 text-center">
-			See also the {getType()}’s <a href={url}>commit history</a>.
-		</p>,
+	const defaultBranch = await getDefaultBranch();
+	if (currentBranch === defaultBranch) {
+		return;
+	}
+
+	// Change branch
+	url.assign({
+		branch: defaultBranch
+	})
+	// Check if that path exists
+	if (await is404(url.toString())) {
+		return;
+	}
+
+	// If it does, return it
+	return url.toString();
+}
+
+async function getLastCommitForFile(
+	branch?: string,
+): Promise<Record<string, string | undefined> | void> {
+	// Get the current branch and file path
+	const url = new GitHubURL(location.href);
+	const currentBranch = url.branch;
+	let filePath = url.filePath;
+
+	// Get the last 2 commits that include the file
+	const commitsForFileResponse = await api.v4(`
+		repository() {
+			ref(qualifiedName: "${branch ?? currentBranch}") {
+				target {
+					... on Commit {
+						history(first: 2, path: "${filePath}") {
+							nodes {
+								oid
+							}
+						}
+					}
+				}
+			}
+		}
+	`);
+	// The latest commit will be the first object in the array
+	const previousCommitsIncludingFile
+		= commitsForFileResponse.repository.ref.target.history.nodes[0];
+	if (!previousCommitsIncludingFile) {
+		return;
+	}
+
+	// Get a list of changes that happened in the repo with this commit
+
+	// API v4 doesn't support retrieving a list of changed files for a commit:
+	// https://github.community/t/graphql-api-get-list-of-files-related-to-commit/14047/2
+	const lastCommitInfo = await api.v3(
+		`commits/${previousCommitsIncludingFile.oid as string}`,
 	);
+
+	// Check what happened to this particular file
+	const fileInfo = lastCommitInfo.files.find((file: AnyObject) => [file.filename, file.previous_filename].includes(filePath));
+	if (!fileInfo) {
+		return;
+	}
+
+	// Get info
+	const commitSha = lastCommitInfo.sha.slice(0, 8);
+	const commitDate = lastCommitInfo.commit.committer.date;
+	const linkToCommit = lastCommitInfo.html_url;
+
+	url.assign({
+		route: 'commits'
+	})
+	const linkToCommitHistory = url.toString();
+
+	if (fileInfo.status === 'removed') {
+		const linkToLastVersion = fileInfo.blob_url;
+
+		return {
+			type: 'removed',
+			filePath,
+			commitSha,
+			commitDate,
+			linkToCommit,
+			linkToLastVersion,
+			linkToCommitHistory,
+		};
+	}
+
+	if (fileInfo.status === 'renamed') {
+		const newFilePath = fileInfo.filename;
```

Comment:
If variables are used once, just declare them where they're used or skip them. For example this one can just be in the returned object:

```diff
 		return {
 			type: 'renamed',
 			filePath,
+ 			newFilePath: fileInfo.filename,
 			commitSha,
 			commitDate,
 			linkToCommit,
 			linkToLastVersion,
 			linkToNewVersion,
 			linkToCommitHistory,
 		};
```

This applies to many other variables

Reply:
Ok, will do